cpp/src/arrow/compute/registry.cc

```diff
@@ -34,78 +34,144 @@ namespace compute {
 
 class FunctionRegistry::FunctionRegistryImpl {
  public:
-  Status AddFunction(std::shared_ptr<Function> function, bool allow_overwrite) {
-#ifndef NDEBUG
-    // This validates docstrings extensively, so don't waste time on it
-    // in release builds.
-    RETURN_NOT_OK(function->Validate());
-#endif
-
-    std::lock_guard<std::mutex> mutation_guard(lock_);
+  explicit FunctionRegistryImpl(FunctionRegistryImpl* parent = NULLPTR)
+      : parent_(parent) {}
+  virtual ~FunctionRegistryImpl() {}
```

Comment:
```suggestion
  ~FunctionRegistryImpl() {}
```
Now that we aren't using inheritance these do not need to be virtual.